### PR TITLE
feat!: Update Questionable Activity Report to support a user defined range of days

### PR DESCRIPTION
**Pull request by tmy1313**
OCD-3199

### CODE REVIEW

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/manager/impl/SchedulerManagerImpl.java [hunk 1]
```diff
@@ -15,6 +15,7 @@
 import java.util.Set;
 
 import org.quartz.CronTrigger;
+import org.quartz.JobDataMap;
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/manager/impl/SchedulerManagerImpl.java]: Should this "Impl" code be put into the "non-Impl" version?

File: chpl/chpl-resources/src/main/resources-production/jobs.xml [hunk 2]
```diff
@@ -27,6 +27,111 @@
                 </entry>
             </job-data-map>
         </job>
+
+        <!-- Questionable URL Data Collector -->
+        <!-- URL Status Data Collector -->
+        <!-- Job used to change status of a set listings -->
+        <job>
```
Comment (andlar) [inline chpl/chpl-resources/src/main/resources-production/jobs.xml]: Why are all of these in this PR? Shouldn't they either be in a separate PR, or if already on staging, not in a PR at all?

File: chpl/chpl-resources/src/main/resources/jobs.xml [hunk 3]
```diff
@@ -394,9 +393,16 @@
                     <key>frequency</key>
                     <value>WEEKLY</value>
                 </entry>
+                <entry>
+                    <key>range</key>
+                    <value>7</value>
+                </entry>
+                <entry>
+                    <key>parameters</key>
+                    <value>[{"label": "Range (in days)", "key": "range", "type": "int", "min": "7"} ]</value>
```
Comment (andlar) [inline chpl/chpl-resources/src/main/resources/jobs.xml]: No specific change request, but why have a min but no max? I thought the ticket talked about a max value?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java [hunk 4]
```diff
@@ -133,6 +134,14 @@ public void execute(final JobExecutionContext jobContext) throws JobExecutionExc
         LOGGER.info("********* Completed the Questionable Activity Email job. *********");
     }
 
+    private Integer getRangeInDays(final JobExecutionContext jobContext) {
+        Integer range = 7;
+        if (jobContext.getMergedJobDataMap().containsKey("range")) {
+            range = jobContext.getMergedJobDataMap().getInt("range");
+        }
+        return range * -1;
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java]: There's no validation in here. I'm not sure if there should be, but we can't assume the UI will provide "valid" numbers. My guess is that we should have a min value and a max value, defined as environment.properties values, and ensure that whatever the end user passes in falls within that range, or is constrained to that range by the code, if we want to allow that

File: chpl/chpl-resources/src/main/resources-production/jobs.xml (hunk 2 above)
Comment (kekey1) [reply chpl/chpl-resources/src/main/resources-production/jobs.xml]: Just to follow up on this - this file as it exists on staging right now has 2 jobs in it.

File: chpl/chpl-resources/src/main/resources/jobs.xml (hunk 3 above)
Comment (tmy1313) [reply chpl/chpl-resources/src/main/resources/jobs.xml]: I think the range should be 1-365.  Updated the code to reflect this.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java (hunk 4 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java]: Updated the code to allow a range of 1-365.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java [hunk 5]
```diff
@@ -41,9 +41,9 @@
 import gov.healthit.chpl.util.Util;
 
 /**
- * The QuestionableActivityEmailJob implements a Quartz job and is available to ROLE_ADMIN. When
- * invoked it emails relevant individuals with the questionable activity that has occurred within
- * the last week.
+ * The QuestionableActivityEmailJob implements a Quartz job and is available to ROLE_ADMIN. When invoked it emails
+ * relevant individuals with the questionable activity that has occurred within the last week.
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java]: Comment is not correct any more

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java [hunk 6]
```diff
@@ -73,9 +73,14 @@
     private static final int ACTIVITY_CERT_STATUS_CHANGE_REASON_COL = 11;
     private static final int ACTIVITY_REASON_COL = 12;
 
+    private static final int MIN_RANGE_IN_DAYS = 1;
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java]: Is there any way to pull these values from the jobs.xml definition? If we ever change them (for whatever reason) we'd have to change them in two places. I'm also thinking about what happens with other jobs that we might add range values for; it'd be nice to have a paradigm for having those values defined in only one place

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java [hunk 7]
```diff
@@ -122,17 +133,33 @@ public void execute(final JobExecutionContext jobContext) throws JobExecutionExc
...
+                    .sendEmail();
 
         } catch (MessagingException e) {
             LOGGER.error(e);
         }
         LOGGER.info("********* Completed the Questionable Activity Email job. *********");
     }
 
+    private Integer getRangeInDays(final JobExecutionContext jobContext) {
+        Integer range = 7;
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java]: This looks like a candidate for a CONSTANT value to use as default if the range is missing

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java [hunk 8]
```diff
@@ -88,9 +93,15 @@ public void execute(final JobExecutionContext jobContext) throws JobExecutionExc
         LOGGER.info("********* Starting the Questionable Activity Email job. *********");
         LOGGER.info("Creating questionable activity email for: " + jobContext.getMergedJobDataMap().getString("email"));
 
+        Integer range = getRangeInDays(jobContext);
+        if (!isRangeValid(range)) {
+            LOGGER.error(
+                    "Range is invalid.  It must be between " + MIN_RANGE_IN_DAYS + " and " + MAX_RANGE_IN_DAYS + ".");
+            return;
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java]: Thinking about the user experience, if someone managed to create a job with an invalid range there's no way they'd ever know it, right? As in, when the job starts it'd have an invalid range and would then just end, without sending any message. Would it make sense to default to the DEFAULT_RANGE value and maybe even add a message to the email about that? We're defaulting them if they don't have a value at all - it might be reasonable to do the same thing if they have an invalid one

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java [hunk 9]
```diff
@@ -689,8 +674,28 @@ private Properties loadProperties() throws IOException {
         return props;
     }
 
-    private void setQuestionableActivityDAO(final QuestionableActivityDAO questionableActivityDao) {
-        this.questionableActivityDao = questionableActivityDao;
+    private void populateRangeDefaultsFromJobData(final JobExecutionContext context) {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/QuestionableActivityEmailJob.java]: What if this function returned a `org.apache.commons.lang3.Range` object, or something similar? Then you could make the current "minRangeInDays and maxRangeInDays actual final static constants, and use the Range object returned from this function in the calling function to do a ".contains(range)" for the actual value

File: chpl/chpl-resources/src/main/resources-production/jobs.xml [hunk 10]
```diff
@@ -25,6 +52,18 @@
...
+                    <key>frequency</key>
+                    <value>WEEKLY</value>
+                </entry>
+                <entry>
+                    <key>range</key>
+                    <value>7</value>
+                </entry>
+                <entry>
+                    <key>parameters</key>
+                    <value>[{"label": "Range (in days)", "key": "range", "type": "int", "min": "1", "max": "365"} ]</value>
```
Comment (andlar) [inline chpl/chpl-resources/src/main/resources-production/jobs.xml]: This job doesn't have these ranges, right?
